File: Altium2RGB.pas
{..............................................................................}
{ Created by Piotr Czapnik 2020                                                }
{..............................................................................}

uses System.Net.HttpClient;

function GetURL(const AURL: string): string;
var
  HttpClient: THttpClient;
  HttpResponse: IHttpResponse;
begin
  HttpClient := THTTPClient.Create;
  try
    HttpResponse := HttpClient.Get(AURL);
    Result := HttpResponse.ContentAsString();
  finally
    HttpClient.Free;
  end;
end;

function getR(RGB: Integer): Byte;
var
  R: Byte;
begin
  R      := (RGB mod $100);
  Result := R;
end;

function getG(RGB: Integer): Byte;
var
  G: Byte;
begin
  G      := ((RGB mod $10000) div $100);
  Result := G;
end;

function getB(RGB: Integer): Byte;
var
  B: Byte;
begin
  B      := RGB div $10000;
  Result := B;
end;


Procedure Altium2RGB;
Var
    PCBBoard     : IPCB_Board;
    color        : TColorRef;
    R            : Byte;
    G            : Byte;
    B            : Byte;
    RRGGBB       : String;
    url          : String;

Begin
    PCBBoard := PCBServer.GetCurrentPCBBoard;
    If PCBBoard = Nil Then Exit;

    color := PCBBoard.LayerColor[PCBBoard.CurrentLayer];
    R     := getR(color);
    G     := getG(color);
    B     := getB(color);
    RRGGBB:= IntToHex(R, 2)+IntToHex(G, 2)+IntToHex(B, 2);
    //url   := 'http://192.168.2.13/cm?cmnd=Color%20' + RRGGBB + '00'; //H801
    url   := 'http://192.168.2.13/json.htm?type=command&param=setcolbrightnessvalue&idx=32&hex=' + RRGGBB; //DOMOTICZ UPDATE DEVICE 32
    //ShowInfo(url);
    GetURL(url);


End;


{..............................................................................}

{..............................................................................}

File: Altium2RGB.js
/*Created by Piotr Czapnik 2020*/
// TODO:
// * ADD HTML REQUEST
// * ADD function event handler if color change

function BGR_to_RRGGBB(BGR) {
var R;
var G;
var B;

    R = Number(BGR&255).toString(16);
    if (R.length < 2) {
       R = "0" + R;
    }

    G = Number((BGR&65280)>>>8).toString(16);
    if (G.length < 2) {
       G = "0" + G;
    }

    B = Number((BGR&16711680)>>>16).toString(16);
    if (B.length < 2) {
       B = "0" + B;
    }

    return R + G + B;
}

function Altium2RGB() {
var Board;            //IPCB_Board;
var RRGGBB;           //hex representation od actual layer color;
var color;
var IP = "192.168.2.13";
var url = "http://" + IP + "/cm?cmnd=Color%20";

    Board = PCBServer.GetCurrentPCBBoard;
    if (Board != Null)
    {
       color = BGR_to_RRGGBB(Board.LayerColor(Board.CurrentLayer));
       url = url + color + "00";

    }
}
